File: app/src/main/java/com/example/taylorshop/MainActivity.java
package com.example.taylorshop;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.SearchView;

import com.example.taylorshop.Adapter.CustomAdapter;
import com.example.taylorshop.Fragment.AddNewCustomer;
import com.example.taylorshop.Models.Customer;
import com.example.taylorshop.interfaces.PopupCallback;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.firebase.FirebaseApp;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.Query;
import com.google.firebase.database.ValueEventListener;
import com.orhanobut.hawk.Hawk;


import java.util.ArrayList;

import dagger.hilt.android.HiltAndroidApp;

@HiltAndroidApp
public class MainActivity extends AppCompatActivity implements PopupCallback {

    public ArrayList<Customer> arrayList = new ArrayList<Customer>();
    FloatingActionButton floatingActionButton;
    public DatabaseReference databaseReference;
    CustomAdapter customAdapter;
    private RecyclerView recyclerView;
    private RecyclerView.Adapter mAdapter;
    private RecyclerView.LayoutManager layoutManager;
    private PopupCallback popupCallback;
    private int position = -1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Hawk.init(this).build();
        popupCallback = this;
        floatingActionButton = findViewById(R.id.fab);
        recyclerView = (RecyclerView) findViewById(R.id.recycler_view);

        floatingActionButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                addCustomer();
            }
        });

        //fetch data offline
        if (!FirebaseApp.getApps(this).isEmpty() && !isOnline())
            if (databaseReference != null)
                FirebaseDatabase.getInstance().setPersistenceEnabled(true);

        databaseReference = FirebaseDatabase.getInstance().getReference().child("Contact");
        databaseReference.keepSynced(true);
        databaseReference.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                arrayList.clear();
                for (DataSnapshot postsnapshot : dataSnapshot.getChildren()) {
                    Customer customer = postsnapshot.getValue(Customer.class);
                    customer.setKey(postsnapshot.getKey());
                    customer.getName();
                    customer.getSerial_number();
                    customer.getPhone_number();
                    arrayList.add(customer);
                }

                mAdapter = new CustomAdapter(arrayList, getApplicationContext(), popupCallback);
                layoutManager = new LinearLayoutManager(MainActivity.this);
                recyclerView.setLayoutManager(layoutManager);
                recyclerView.setAdapter(mAdapter);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu, menu);

        MenuItem menuItem = menu.findItem(R.id.action_search);
        SearchView searchView = (SearchView) menuItem.getActionView();
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String s) {
                return false;
            }

            @Override
            public boolean onQueryTextChange(String s) {
                filter(s);
                return true;
            }
        });

        return true;
    }

    public void filter(String text) {
        ArrayList<Customer> srchList = new ArrayList<Customer>();
        if (text.isEmpty()) {
            srchList.addAll(arrayList);
            Log.e("Customer", String.valueOf(arrayList.size()));
        } else {
            text = text.toLowerCase();
            for (Customer item : arrayList) {
                if (item.getName().toLowerCase().contains(text) || item.getPhone_number().toLowerCase().contains(text)) {
                    srchList.add(item);
                }
            }
        }
        mAdapter = null;
        mAdapter = new CustomAdapter(srchList, getApplicationContext());
        recyclerView.setAdapter(mAdapter);
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
    }

    public void addCustomer() {
        Fragment fragment = new AddNewCustomer();
        FragmentTransaction fragmentTransaction = getSupportFragmentManager().beginTransaction();
        fragmentTransaction.replace(R.id.frame, fragment, "MyFragment");
        fragmentTransaction.addToBackStack("Customer");
        fragmentTransaction.commit();
    }

    public boolean isOnline() {
        ConnectivityManager connMgr = (ConnectivityManager)
                getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo networkInfo = connMgr.getActiveNetworkInfo();
        return (networkInfo != null && networkInfo.isConnected());
    }

    @Override
    public void onClick(int position) {
        DatabaseReference ref = FirebaseDatabase.getInstance().getReference();
        Query sQuery = ref.child("Contact").orderByChild("phone_number").equalTo(arrayList.get(position).getPhone_number());
        sQuery.addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                for (DataSnapshot snapshot : dataSnapshot.getChildren()) {
                    snapshot.getRef().removeValue();
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
        arrayList.remove(position);
        arrayList.clear();
        if (customAdapter != null) customAdapter.notifyDataSetChanged();
    }
}

File: app/src/main/java/com/example/taylorshop/MainActivity.kt
package com.example.taylorshop

import android.annotation.SuppressLint
import android.net.ConnectivityManager
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.View
import android.widget.SearchView
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.taylorshop.Adapter.CustomAdapter
import com.example.taylorshop.Fragment.AddNewCustomer
import com.example.taylorshop.Models.Customer
import com.example.taylorshop.data.AppDatabase
import com.example.taylorshop.interfaces.PopupCallback
import com.example.taylorshop.utilities.CustomerVM
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.firebase.database.*
import com.orhanobut.hawk.Hawk

class MainActivity : AppCompatActivity(), PopupCallback {
    lateinit var databaseReference: DatabaseReference
    var arrayList = ArrayList<Customer?>()
    private lateinit var floatingActionButton: FloatingActionButton
    private lateinit var recyclerView: RecyclerView
    private var mAdapter: CustomAdapter? = null
    private var popupCallback: PopupCallback? = null
    private lateinit var customerVM: CustomerVM

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        Hawk.init(this).build()
        popupCallback = this
        floatingActionButton = findViewById(R.id.fab)
        customerVM = ViewModelProvider(this)[CustomerVM::class.java]
        recyclerView = findViewById(R.id.recycler_view)
        floatingActionButton.setOnClickListener(View.OnClickListener { addCustomer() })

        recyclerView.layoutManager = LinearLayoutManager(this)
        databaseReference = FirebaseDatabase.getInstance().reference.child("Contact")
        if (!Hawk.contains("Fetched_from_firebase")) {
            databaseReference.addListenerForSingleValueEvent(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    arrayList.clear()
                    for (postsnapshot in dataSnapshot.children) {
                        Hawk.put("Fetched_from_firebase", true)
                        val customer = postsnapshot.getValue(Customer::class.java)
                        customer!!.key = postsnapshot.key
                        arrayList.add(customer)
                        Hawk.put("userID", customer.id)
                        AppDatabase.getInstance(applicationContext).customerDao().insertCustomer(customer)
                    }
                }
                override fun onCancelled(databaseError: DatabaseError) {}
            })
        }

        if (isOnline) customerVM.deleteAfterOnline(databaseReference)

            customerVM.customerList.observe(this) {
                if (!it.isNullOrEmpty()) {
                    arrayList.clear()
                    customerVM.checkInternetConnectivity(isOnline,this, databaseReference, it)
                    arrayList.addAll(it as ArrayList<Customer>)
                    mAdapter = CustomAdapter(arrayList, applicationContext, popupCallback)
                    recyclerView.adapter = mAdapter
                } else {
                    customerVM.checkInternetConnectivity(isOnline,this, databaseReference, it)
                }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu, menu)
        val menuItem = menu.findItem(R.id.action_search)
        val searchView = menuItem.actionView as SearchView
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(s: String): Boolean {
                return false
            }

            override fun onQueryTextChange(s: String): Boolean {
                filter(s)
                return true
            }
        })
        return true
    }

    fun filter(text: String) {
        var text = text
        val srchList = ArrayList<Customer?>()
        if (text.isEmpty()) {
            srchList.addAll(arrayList)
            Log.e("Customer", arrayList.size.toString())
        } else {
            text = text.toLowerCase()
            for (item in arrayList) {
                if (item!!.name!!.toLowerCase().contains(text) || item.phone_number!!.toLowerCase()
                        .contains(text)
                ) {
                    srchList.add(item)
                }
            }
        }
        mAdapter = null
        mAdapter = CustomAdapter(srchList, applicationContext)
        recyclerView.adapter = mAdapter
    }

    private fun addCustomer() {
        val fragment: Fragment = AddNewCustomer()
        val fragmentTransaction = supportFragmentManager.beginTransaction()
        fragmentTransaction.replace(R.id.frame, fragment, "MyFragment")
        fragmentTransaction.addToBackStack("Customer")
        fragmentTransaction.commit()
    }

    private val isOnline: Boolean
        get() {
            val connMgr = getSystemService(CONNECTIVITY_SERVICE) as ConnectivityManager
            @SuppressLint("MissingPermission") val networkInfo = connMgr.activeNetworkInfo
            return networkInfo != null && networkInfo.isConnected
           }

    override fun onClick(position: Int) {
         AppDatabase.getInstance(this).customerDao().deleteCustomer(arrayList[position]!!)
        if(isOnline) {
            val sQuery = FirebaseDatabase.getInstance().reference.child("Contact").orderByChild("phone_number")
                    .equalTo(arrayList[position]?.phone_number)
            sQuery.addListenerForSingleValueEvent(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    for (snapshot in dataSnapshot.children) {
                        snapshot.ref.removeValue()
                    }
                }

                override fun onCancelled(databaseError: DatabaseError) {}
            })
        } else {
            if (Hawk.contains("delete_customer")) {
                val data = Hawk.get<HashMap<String, Customer>>("delete_customer")
                data[arrayList[position]?.phone_number.toString()] = arrayList[position]!!
                Hawk.put("delete_customer", data)
            } else {
                val deleteUsers = HashMap<String, Customer>()
                deleteUsers[arrayList[position]?.phone_number.toString()] = arrayList[position]!!
                Hawk.put("delete_customer" , deleteUsers)
            }
        }
        if (arrayList.size == 1 && position == 0) arrayList.clear() else arrayList.removeAt(position)
        mAdapter?.notifyItemRemoved(position)
    }
}